evm_arithmetization/src/prover.rs

```diff
@@ -526,6 +527,11 @@ pub struct SegmentDataIterator<F: RichField> {
     partial_next_data: Option<GenerationSegmentData>,
 }
 
+pub enum SegmentRunResult {
```

Comment:
Fair enough. I'll probably keep the name as associated type for simplicity.

Reply:
is it something like

```
pub type SegmentRunResult = Option<(GenerationSegmentData, Option<GenerationSegmentData>)>
```
?

Reply:
Now I'm confused, I thought you didn't want `generate_next_segment` to return an `Option` 😅

Reply:
As discussed offline, I think `generate_next_segment` should return Ok<_> if the execution is regular, and Err(_) if it is irregular. If the execution is regular, we return Option<_> to mark if there is more segments, or we have finished producing segments.
For the `Ok` variant then it could be (removing `Result` from name to prevent confusion):
```
pub type SegmentRunData = Option<(GenerationSegmentData, Option<GenerationSegmentData>)>
```.
This way `generate_next_segment` clearly communicates what happened. As we need to return then from iterator, we restructure and convert to Some(Result) or None.